core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,155 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.gateway.MetaStateService;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexSettings;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+
+/**
+ * Renames index folders from {index.name} to {index.uuid}
+ */
+public class IndexFolderUpgrader {
+    private final NodeEnvironment nodeEnv;
+    private final Settings settings;
+    private final ESLogger logger = Loggers.getLogger(IndexFolderUpgrader.class);
+    private final MetaDataStateFormat<IndexMetaData> indexStateFormat = metaDataMetaDataStateFormat();
+
+    /**
+     * Creates a new upgrader instance
+     * @param settings node settings
+     * @param nodeEnv the node env to operate on
+     */
+    IndexFolderUpgrader(Settings settings, NodeEnvironment nodeEnv) {
+        this.settings = settings;
+        this.nodeEnv = nodeEnv;
+    }
+
+    /**
+     * Renames the index folder found in <code>source</code>
+     * if the renamed folder does not exist, otherwise no-op
+     */
+    void upgrade(final Index index, final Path source) throws IOException {
+        final Path target = source.resolveSibling(index.getUUID());
+        if (Files.exists(target)) {
+            // skip we upgraded already
+            // we assume no not-yet-upgraded file named {index_uuid} exists
+            logger.info("[{}] no upgrade needed - already upgraded", target);
+            return;
+        }
+        boolean success = false;
+        try {
+            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
+            success = true;
+        } catch (NoSuchFileException ignored) {
+            assert ignored.getFile().equals(source.toAbsolutePath().toString()) : "must be same as the source";
+            // thrown when the source is non-existent because the folder was renamed
+            // by another node (shared FS) after we checked if the target exists
+            assert Files.exists(target) : "must be upgraded by another node";
```

Comment:
I removed support for this and documented it in the migration guide. Now we throw the exception and log an error message to retry upgrading when we detect `NoSuchFileException` or `FileNotFoundException`